lib/SILGen/SILGenExpr.cpp

```diff
@@ -2293,6 +2293,20 @@ RValue RValueEmitter::visitRebindSelfInConstructorExpr(
   return SGF.emitEmptyTupleRValue(E, C);
 }
 
+static bool isNullableTypeInC(Type ty) {
+  if (ty->hasReferenceSemantics())
+    return true;
+  
+  auto &C = ty->getASTContext();
+  auto nom = ty->getAnyNominal();
+  if (!nom)
+    return false;
+
+  return nom == C.getUnsafePointerDecl()
+    || nom == C.getUnsafeMutablePointerDecl()
+    || nom == C.getAutoreleasingUnsafeMutablePointerDecl();
+}
```

Comment:
Using `hasReferenceSemantics` to cover (thin) function pointers also feels a little sketchy, but I suppose we're not ever going to change that.